### PR TITLE
Feat/thread sleep

### PR DESCRIPTION
**Pull request by mckeenicholas**
Adds thread sleep and wakeup functions and a mutex using a sleeping lock.

### CODE REVIEW

File: kernel/src/sync/mutex/sleep.rs
Comment (ff-zhang) [inline kernel/src/sync/mutex/sleep.rs:1]: If we add a `mutex` folder, should we move `mutex.rs` into it?

File: kernel/src/threading/scheduling/scheduler.rs [hunk 1]
```diff
@@ -10,4 +10,5 @@ pub trait Scheduler {
     fn push(&mut self, thread: Box<ThreadControlBlock>);
     fn pop(&mut self) -> Option<Box<ThreadControlBlock>>;
     fn remove(&mut self, tid: Tid) -> Option<Box<ThreadControlBlock>>;
+    fn get(&mut self, tid: Tid) -> Option<&mut ThreadControlBlock>;
```
Comment (ff-zhang) [inline kernel/src/threading/scheduling/scheduler.rs:13]: In Rust, I think this should be name `get_mut`

**Review by ff-zhang — APPROVED** [review]
LGTM but I didn't test `SleepMutex`; it would be a good idea to wait for someone to review before merging

File: kernel/src/sync/mutex/sleep.rs [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+            thread_wakeup(next_thread);
+        }
+
+        self.aquired = false;
+        self.holding_thread = None;
+        intr_enable();
+    }
+
+    pub fn is_locked(&mut self) -> bool {
+        self.aquired
```
Comment (1whatleytay) [inline kernel/src/sync/mutex/sleep.rs:116]: `&mut self`? Is this just for exclusivity?

**Review by 1whatleytay — APPROVED** [review]

**Pull request merged**